include/openssl/core_numbers.h

```diff
@@ -57,6 +58,12 @@ OSSL_CORE_MAKE_FUNC(const OSSL_ITEM *,
 # define OSSL_FUNC_CORE_GET_PARAMS             2
 OSSL_CORE_MAKE_FUNC(int,core_get_params,(const OSSL_PROVIDER *prov,
                                          const OSSL_PARAM params[]))
+# define OSSL_FUNC_CORE_PUT_ERROR              3
+OSSL_CORE_MAKE_FUNC(void,core_put_error,(int lib, int func, int reason,
+                                         const char *file, int line))
+# define OSSL_FUNC_CORE_ADD_ERROR_VDATA        4
+OSSL_CORE_MAKE_FUNC(void,core_add_error_vdata,(int num, va_list args))
+
```

Comment:
I've been staring at this many times, thinking something is missing...  and yeah, you want upcalls for `ERR_func_error_string` and `ERR_load_strings_const` as well, for the sake of `ERR_load_PROV_strings`

Reply:
I think that can go in a separate PR if needed.  I don't think the FIPS provider will add error codes that aren't in the core, and having providers add their own error codes is a bigger problem that would "distract" from the main point of this.

Reply:
> I've been staring at this many times, thinking something is missing... and yeah, you want upcalls for ERR_func_error_string and ERR_load_strings_const as well, for the sake of ERR_load_PROV_strings

Why do we need that? Error string are all libcrypto side - why do we need to expose them to the provider?

Reply:
Uhmmm, it's the other way around.  Provider side error messages should be in the provider, and the means that the provider will need to have them loaded so libcrypto can display them right.

Reply:
Providers cannot load their own error messages until OpenSSL defines a way for an arbitrary third-party to get a library/facility that doesn't conflict with any other. As the number of bits available is very very small, solving this will be a hard problem.

Reply:
The number of bits is 8, allowing up to 256 different error "modules". There is actually a function to obtain the next available number: `ERR_get_next_error_library`. That one would need an upcall too. 

Note that this is with our current error model. That one could be remade to become much simpler to handle.

Reply:
> Uhmmm, it's the other way around. Provider side error messages should be in the provider, and the means that the provider will need to have them loaded so libcrypto can display them right.

We need a new error system for the providers. We need to give some thought into how that will work. This PR isn't the right location to have that discussion. All I'm doing here is the minimum necessary to get this to work for now. It is not a long term solution.

I fully expect the long term solution to have errors on the provider side. I also expect that these errors will work differently to the way the current error system works. But we are not there yet. Right now we are using the existing error system. Right now error strings are loaded on the libcrypto side. Exposing upcalls for `ERR_func_error_string`, `ERR_load_strings_const`and `ERR_load_PROV_strings` is not needed for this PR. Because I expect the new error system to work differently to the current system I do not believe we should be anticipating what upcalls we are going to need for that new error system in this PR. 

I'm only exposing this particular upcall because it is necessary to get things to work right now. I do expect this to change.

Reply:
I can accept that.

Reply:
Another option, *for this release,* is to not solve this problem, and instead do what I provide in #8887

Reply:
#8887, in *addition* to what is done in this PR, makes `ERR_add_error_vdata` append strings instead of replacing (which I think is a good thing, don't get me wrong). That could be an independent PR.

Reply:
Yes, that is what I meant by "how I addressed the issue." :)  If there are concerns about that change, I will make a new API to append a text string. But that should be discussed over in #8887.

This PR needs the set-error API's, and @mattcaswell didn't want to delay and rebase again for 8887 to go in first.